### DISCUSSION TITLE
Going to deep sleep even sleep cylce set to 0 seconds

### DISCUSSION THREAD

**Discussion by t-huyeng** [cyberman54/ESP32-Paxcounter#899] [Q&A]
I wonder why the my board goes to sleep after I set the sleep cycle to 0 seconds.
```
11:11:58.650 > [ 61343][I][rcommand.cpp:89] set_sleepcycle(): [src/rcommand.cpp] Remote command: set sleep cycle to 0 seconds
11:11:58.659 > [ 61350][I][rcommand.cpp:407] set_saveconfig(): [src/rcommand.cpp] Remote command: save config to NVRAM
11:11:58.668 > [ 61359][I][configmanager.cpp:69] saveConfig(): [src/configmanager.cpp] Storing settings to NVRAM...
11:11:58.677 > [ 61375][I][configmanager.cpp:85] saveConfig(): [src/configmanager.cpp] Device settings saved
11:11:58.841 > [ 61543][I][lorawan.cpp:219] lora_send(): [src/lorawan.cpp] 3 byte(s) sent to LORA
11:11:59.562 > [ 62264][I][reset.cpp:119] enter_deepsleep(): [src/reset.cpp] Waiting until LMIC is idle...
11:12:05.562 > [ 68264][I][lorawan.cpp:532] SaveLMICToRTC(): [src/lorawan.cpp] LMIC state saved
11:12:05.570 > [ 68266][I][reset.cpp:179] enter_deepsleep(): [src/reset.cpp] Going to sleep, good bye.
```

using: `halfile = heltecv2.h` current master but with `espressif32@5.1.0`

If no sleep cycle is set before sending this remote command, it look like the esp32 does not wake up at all.

**Comment by cyberman54** [comment]
current code needs espressif32@5.2.0, it won't work with 5.1.0. This affects all your reported issues.

**Reply by t-huyeng** [reply]
I got `espressif32@5.2.0`  working with a fresh install, but as far as I can see it at the moment the issues are at least for me not solved. I will try to investigate this further.

**Reply by cyberman54** [reply]
@t-huyeng 
Please check this scenario, which could explain this issue:
1) You set a default sleepcycle X > 0in paxcounter.conf, you've build code with this setting and flashed it to your board
2) The board will use X as default when powering up
3) During runtime you changed X = 0 by rcommand, but you did not make this setting permanent by rcommand 0x21 (see Readme.md)
4) You reset the module
5) During powering up the module will reload the default setting 1) from NVRAM

To avoid this, you must add a 0x21 to your rcommand in step 3) to make setting permanent.

**Reply by t-huyeng** [reply]
I tested it again and I do not think this is the problem: 
the logs also show that the setting is stored to the NVRAM
```
11:11:58.650 > [ 61343][I][rcommand.cpp:89] set_sleepcycle(): [src/rcommand.cpp] Remote command: set sleep cycle to 0 seconds
11:11:58.659 > [ 61350][I][rcommand.cpp:407] set_saveconfig(): [src/rcommand.cpp] Remote command: save config to NVRAM
11:11:58.668 > [ 61359][I][configmanager.cpp:69] saveConfig(): [src/configmanager.cpp] Storing settings to NVRAM...
```
I am just unsure why the board goes into deep sleep at all, when I set it to 0 seconds. Does it mean the board will always do one more round of sleeping with the old settings before applying the new sleep cycle? (I tested also with other values than 0 and for me it looks like that) 
If this is the case I think it might be helpful to change it, so that the new sleep cycle time is used directly after it has been send to the board. 
Until now I could not find the right spot in the code but I will look again.

**Reply by cyberman54** [reply]
Please try v3.4.3 in development branch. I fixed some oddities with sleep mode.

**Reply by cyberman54** [reply]
>I am just unsure why the board goes into deep sleep at all, when I set it to 0 seconds. Does it mean the board will always do one more >round of sleeping with the old settings before applying the new sleep cycle? (I tested also with other values than 0 and for me it looks >like that)
>If this is the case I think it might be helpful to change it, so that the new sleep cycle time is used directly after it has been send to the >board.

@t-huyeng That's true, it takes one more cycle. This is because the rcommand with sleep setting arrives AFTER the payload is sent and going to sleep is already ongoging. I will take a look into code it this could be changed without additional complexity by a queue, etc.

**Reply by t-huyeng** [reply]
Thanks! 
I think this is probably also what your wrote in the other [discussion: ](https://github.com/cyberman54/ESP32-Paxcounter/discussions/896#discussioncomment-4056076)

>The high level logic for triggering a sleep is as follows:
> - check if a sleep cycle is configured happenes after each senddata cycle (purpose is the device goes to sleep after data of a cycle is sent, not before)
> - senddata is triggered either by libpax lib callback (if wifi and/or ble count is active) OR by xTimer, if libpax lib is not active (devices without wifi/ble count).


For me it would make more sense, if the "new" sleep cycle is used directly after receiving. I will also try to look into the code, but I am sure you are more familiar and know where to look. If I can help let me know. I will for now check v3.4.3 code.
